// tdd-guard/dist/config/Config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = exports.DEFAULT_DATA_DIR = exports.DEFAULT_CLIENT = exports.DEFAULT_MODEL_VERSION = void 0;
const path_1 = __importDefault(require("path"));
const TEST_RESULTS_FILENAME = 'test.json';
const TODOS_FILENAME = 'todos.json';
const MODIFICATIONS_FILENAME = 'modifications.json';
const LINT_FILENAME = 'lint.json';
const CONFIG_FILENAME = 'config.json';
const INSTRUCTIONS_FILENAME = 'instructions.md';
exports.DEFAULT_MODEL_VERSION = 'claude-sonnet-4-0';
exports.DEFAULT_CLIENT = 'sdk';
exports.DEFAULT_DATA_DIR = path_1.default.join('.claude', 'tdd-guard', 'data');
const VALID_CLIENTS = new Set(['api', 'cli', 'sdk']);
const MODEL_TYPE_TO_CLIENT = {
    anthropic_api: 'api',
    claude_cli: 'cli',
};
class Config {
    dataDir;
    useSystemClaude;
    anthropicApiKey;
    modelType;
    linterType;
    modelVersion;
    validationClient;
    constructor(options) {
        const mode = options?.mode ?? 'production';
        this.dataDir = this.getDataDir(options);
        this.useSystemClaude = this.getUseSystemClaude(options);
        this.anthropicApiKey = this.getAnthropicApiKey(options);
        this.modelType = this.getModelType(options, mode);
        this.linterType = this.getLinterType(options);
        this.modelVersion = this.getModelVersion(options);
        this.validationClient = this.getValidationClient(options);
    }
    getDataDir(options) {
        // Determine the base directory
        const baseDir = options?.projectRoot ?? this.getValidatedClaudeProjectDir();
        // If we have a base directory, construct the full path
        if (baseDir) {
            return path_1.default.join(baseDir, exports.DEFAULT_DATA_DIR);
        }
        // Default to relative path
        return exports.DEFAULT_DATA_DIR;
    }
    getUseSystemClaude(options) {
        return options?.useSystemClaude ?? process.env.USE_SYSTEM_CLAUDE === 'true';
    }
    getAnthropicApiKey(options) {
        return options?.anthropicApiKey ?? process.env.TDD_GUARD_ANTHROPIC_API_KEY;
    }
    getModelType(options, mode) {
        return (options?.modelType ?? this.getEnvironmentModelType(mode) ?? 'claude_cli');
    }
    getEnvironmentModelType(mode) {
        if (mode === 'test' && process.env.TEST_MODEL_TYPE) {
            return process.env.TEST_MODEL_TYPE;
        }
        return process.env.MODEL_TYPE;
    }
    get testResultsFilePath() {
        return path_1.default.join(this.dataDir, TEST_RESULTS_FILENAME);
    }
    get todosFilePath() {
        return path_1.default.join(this.dataDir, TODOS_FILENAME);
    }
    get modificationsFilePath() {
        return path_1.default.join(this.dataDir, MODIFICATIONS_FILENAME);
    }
    get lintFilePath() {
        return path_1.default.join(this.dataDir, LINT_FILENAME);
    }
    get configFilePath() {
        return path_1.default.join(this.dataDir, CONFIG_FILENAME);
    }
    get instructionsFilePath() {
        return path_1.default.join(this.dataDir, INSTRUCTIONS_FILENAME);
    }
    getLinterType(options) {
        if (options && 'linterType' in options) {
            return options.linterType?.toLowerCase();
        }
        const envValue = process.env.LINTER_TYPE?.toLowerCase();
        return envValue && envValue.trim() !== '' ? envValue : undefined;
    }
    getModelVersion(options) {
        return (options?.modelVersion ??
            process.env.TDD_GUARD_MODEL_VERSION ??
            exports.DEFAULT_MODEL_VERSION);
    }
    getValidationClient(options) {
        return (getClientType(options) ?? getLegacyClientType(options) ?? exports.DEFAULT_CLIENT);
    }
    getValidatedClaudeProjectDir() {
        const projectDir = process.env.CLAUDE_PROJECT_DIR;
        if (!projectDir) {
            return null;
        }
        // Validate that CLAUDE_PROJECT_DIR is an absolute path
        if (!path_1.default.isAbsolute(projectDir)) {
            throw new Error('CLAUDE_PROJECT_DIR must be an absolute path');
        }
        // Validate that CLAUDE_PROJECT_DIR does not contain path traversal
        if (projectDir.includes('..')) {
            throw new Error('CLAUDE_PROJECT_DIR must not contain path traversal');
        }
        // Validate that current working directory is within CLAUDE_PROJECT_DIR
        const cwd = process.cwd();
        if (!cwd.startsWith(projectDir)) {
            throw new Error('CLAUDE_PROJECT_DIR must contain the current working directory');
        }
        return projectDir;
    }
}
exports.Config = Config;
function getClientType(options) {
    const directValue = options?.validationClient ?? process.env.VALIDATION_CLIENT;
    const normalizedValue = directValue?.toLowerCase();
    return isValidClient(normalizedValue) ? normalizedValue : undefined;
}
function getLegacyClientType(options) {
    const modelType = options?.modelType ?? process.env.MODEL_TYPE;
    return mapModelTypeToClient(modelType);
}
function isValidClient(value) {
    return VALID_CLIENTS.has(value ?? '');
}
function mapModelTypeToClient(modelType) {
    const normalizedType = modelType?.toLowerCase();
    return normalizedType ? MODEL_TYPE_TO_CLIENT[normalizedType] : undefined;
}
